### ISSUE TITLE
Defining API Standards with Web Concepts

### ISSUE THREAD

**Issue by smizell** [dret/sedola#52]
I wanted to share a little idea I had a while back that feels inline with the spirit of this project. It came up as we were defining API standards at work. My observations were:

1. **We rarely used an entire standard**. For instance, for conditional requests, we said that `Etag` was a MUST and last modified was a MAY.
1. **The standards were made to be broader than what was used in APIs**. Back to my previous example, the API standards internally said MUST for `Etag` and MAY for last modified, but an API might only use the `Etag` and not list the last modified so client developers didn't have to support it.

We were providing company standards while also letting APIs share which of those standards they were using. After I sat on in many of the discussions, I had the thought of some format for describing company standards. 

https://gist.github.com/smizell/d3419fcb358d057807cc18d8898d94b2

To sum up then, I have found that it's not enough to say what standards or concepts are used when defining some internal standards. People need to define a subset of what they allow and a subset of that of what a particular API used. Maybe then there is an opportunity to be able to narrow down concepts and standards in this way.

**Comment by dret** [comment]
On 2019-06-03 17:26, Stephen Mizell wrote:
> To sum up then, I have found that it's not enough to say what standards 
> or concepts are used when defining some internal standards. People need 
> to define a subset of what they allow and a subset of that of what a 
> particular API used. Maybe then there is an opportunity to be able to 
> narrow down concepts and standards in this way.

thanks, that's an interesting observation. my very first draft actually 
had a "description" member in there to describe how the concept is used. 
would that be something you would find useful? an optional member for a 
human-readable description of the concept's use in the API?

generally speaking, please keep in mind that web concepts works on the 
level of individual concept values (e.g., header field types). standards 
may define a whole set of those, but at the level of web concepts, you 
don't see the standards directly (unless you dive into Web Concepts data 
itself, which you can), just the concepts you're using:

for example, "http-header" and "ETag" would identify the header field 
type you used as an example. it is described at 
http://webconcepts.info/http-header/ETag, which links to 
http://webconcepts.info/specs/IETF/RFC/7232, and only there you'd find 
the "Last-Modified" header field type and would be able to tell that RFC 
7232 defining "ETag" defines more concepts than just this one header 
field type.

**Comment by smizell** [comment]
I think I see it how you are describing things. People are interested in using specific concepts from larger standards and not necessarily the standard as a whole. For example, to narrow my example even further, someone may support `Etag` only on `GET` requests but not use it for `PUT` or `DELETE` for whatever reason. To some, that may not be useful, but it was part of the conversation. I think for some, they don't to be specific so clients know what to expect.

So in a situation like this, a company-wide standard may say "You MAY use Etag for GET, PUT, DELETE, etc." Then an individual API may say it may use Etag only for GET.

The reason why I did the example format the way I did it was because there is also the question API developers will have about what all the "MUSTs" are for an implementation. We have a bunch of markdown documents, so finding this out programmatically will be difficult. I'd love to have a checklist of all the concepts/standards/guidelines needed.

**Comment by dret** [comment]
On 2019-06-03 13:58, Stephen Mizell wrote:
> I think I see it how you are describing things. People are interested in 
> using specific concepts from larger standards and not necessarily the 
> standard as a whole. For example, to narrow my example even further, 
> someone may support |Etag| only on |GET| requests but not use it for 
> |PUT| or |DELETE| for whatever reason. To some, that may not be useful, 
> but it was part of the conversation. I think for some, they don't to be 
> specific so clients know what to expect.

the goal of webconcepts isn't to capture every single constraint that 
people might choose to define. it is simply to give them a way to 
identify and document the topics that they use. if you want to find out 
*exactly* how they are used, in many cases you will have to go to the 
documentation and find out all the details.

> So in a situation like this, a company-wide standard may say "You MAY 
> use Etag for GET, PUT, DELETE, etc." Then an individual API may say it 
> may use Etag only for GET.

well, i'd simply say that 
http://webconcepts.info/concepts/http-header/ETag is part of the 
vocabulary that APIs may use, and then they use it in whatever way works 
best for them.

> The reason why I did the example format the way I did it was because 
> there is also the question API developers will have about what all the 
> "MUSTs" are for an implementation. We have a bunch of markdown 
> documents, so finding this out programmatically will be difficult. I'd 
> love to have a checklist of all the concepts/standards/guidelines needed.

i am not sure webconcepts will be a big help for MUSTs. it's mostly 
meant to give you an easy way to describe the elements that you are 
using to compose your APIs. the specific way in which you are composing 
it is outside the scope of webconcepts. so maybe it could be useful as a 
starting point for you, but that may be all.

**Comment by smizell** [comment]
> the specific way in which you are composing it is outside the scope of webconcepts. so maybe it could be useful as a starting point for you, but that may be all.

That makes sense. As you describe it, I can see it as a building block for standards. I'll explore that idea in the future. Thanks for taking the time to share your thoughts on this.
